standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java

```diff
@@ -0,0 +1,311 @@
+  public static final String JDBC_PARTITION_COLUMN = JDBC_CONFIG_PREFIX + ".partitionColumn";
+  public static final String JDBC_NUM_PARTITIONS = JDBC_CONFIG_PREFIX + ".numPartitions";
+  public static final String JDBC_LOW_BOUND = JDBC_CONFIG_PREFIX + ".lowerBound";
+  public static final String JDBC_UPPER_BOUND = JDBC_CONFIG_PREFIX + ".upperBound";
+
+  private static final String JDBC_INPUTFORMAT_CLASS = "org.apache.hive.storage.jdbc.JdbcInputFormat".intern();
+  private static final String JDBC_OUTPUTFORMAT_CLASS = "org.apache.hive.storage.jdbc.JdbcOutputFormat".intern();
+
+  String type = null; // MYSQL, POSTGRES, ORACLE, DERBY, MSSQL, DB2 etc.
+  String driverClassName = null;
+  String jdbcUrl = null;
+  String username = null;
+  String password = null; // TODO convert to byte array
+
+  public AbstractJDBCConnectorProvider(String dbName, DataConnector dataConn) {
+    super(dbName, dataConn);
+    this.type = connector.getType().toUpperCase(); // TODO
+    this.jdbcUrl = connector.getUrl();
+    this.username = connector.getParameters().get(JDBC_USERNAME);
+    this.password = connector.getParameters().get(JDBC_PASSWORD);
+    if (this.password == null) {
+      String keystore = connector.getParameters().get(JDBC_KEYSTORE);
+      String key = connector.getParameters().get(JDBC_KEY);
+      try {
+        char[] keyValue = MetastoreConf.getValueFromKeystore(keystore, key);
+        if (keyValue != null)
+          this.password = new String(keyValue);
+      } catch (IOException i) {
+        LOG.warn("Could not read key value from keystore");
+      }
+    }
+
+    try {
+      warehouse = new Warehouse(MetastoreConf.newMetastoreConf());
+    } catch (MetaException e) { /* ignore */ }
+  }
+
+  @Override public void open() throws ConnectException {
+    try {
+      Class.forName(driverClassName);
+      handle = DriverManager.getConnection(jdbcUrl, username, password);
+      isOpen = true;
+    } catch (ClassNotFoundException cnfe) {
+      LOG.warn("Driver class not found in classpath:" + driverClassName);
+      throw new RuntimeException("Driver class not found:" + driverClassName);
+    } catch (SQLException sqle) {
+      LOG.warn("Could not connect to remote data source at " + jdbcUrl);
+      throw new ConnectException("Could not connect to remote datasource at " + jdbcUrl + ",cause:" + sqle.getMessage());
+    }
+  }
+
+  protected Connection getConnection() {
+    try {
+      if (!isOpen)
+        open();
+    } catch (ConnectException ce) {
+      throw new RuntimeException(ce.getMessage());
+    }
+
+    if (handle instanceof Connection)
+      return (Connection)handle;
+
+    throw new RuntimeException("unexpected type for connection handle");
+  }
+
+  @Override public void close() {
+    if (isOpen) {
+      try {
+        ((Connection)handle).close();
+      } catch (SQLException sqle) {
+        LOG.warn("Could not close jdbc connection to " + jdbcUrl, sqle);
```

Comment:
Should we throw the SQLException forward and let it be handles in the upper layers?

Reply:
throwing a new RuntimeException with the underlying exception. FIXED.